From 90818bc8a036d7a8633acb6e6352ec76d1bcbe0e Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Mon, 3 Oct 2022 17:12:50 +0300
Subject: [PATCH 2/6] serial: linflex: Get Linflex clock rate during
 initialization

Avoid multiple calls of 'clock_get_rate' because some of them
might land in other components (ATF, SCP) which use the same
Linflex instance.
In some cases this might block the execution because the clock
request uses an uninitialized Linflex module.

Issue: ALB-9298
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/serial/serial_linflexuart.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/serial/serial_linflexuart.c b/drivers/serial/serial_linflexuart.c
index 893ecc298f..004bb6b833 100644
--- a/drivers/serial/serial_linflexuart.c
+++ b/drivers/serial/serial_linflexuart.c
@@ -168,6 +168,7 @@ static int _linflex_serial_init(struct linflex_fsl *base, ulong rate)
 struct linflex_serial_priv {
 	struct linflex_fsl *lfuart;
 	struct clk clk;
+	unsigned long rate;
 };
 
 int linflex_serial_setbrg(struct udevice *dev, int baudrate)
@@ -175,7 +176,7 @@ int linflex_serial_setbrg(struct udevice *dev, int baudrate)
 	struct linflex_serial_priv *priv = dev_get_priv(dev);
 
 	_linflex_enter_init(priv->lfuart);
-	_linflex_serial_setbrg(priv->lfuart, clk_get_rate(&priv->clk),
+	_linflex_serial_setbrg(priv->lfuart, priv->rate,
 			       baudrate);
 	_linflex_exit_init(priv->lfuart);
 
@@ -214,7 +215,6 @@ static int linflex_serial_probe(struct udevice *dev)
 {
 	struct linflex_serial_priv *priv = dev_get_priv(dev);
 	fdt_addr_t base_addr;
-	ulong rate;
 	const char *clk_name = "lin";
 	int ret;
 
@@ -234,10 +234,10 @@ static int linflex_serial_probe(struct udevice *dev)
 		return ret;
 	}
 
-	rate = clk_get_rate(&priv->clk);
+	priv->rate = clk_get_rate(&priv->clk);
 
 	priv->lfuart = (struct linflex_fsl *)base_addr;
-	_linflex_serial_init(priv->lfuart, rate);
+	_linflex_serial_init(priv->lfuart, priv->rate);
 
 	return 0;
 }
-- 
2.17.1

